src/v/kafka/client/transport.h

```diff
@@ -135,51 +135,49 @@ class transport : public net::base_transport {
     /*
      * Invokes dispatch with the given request type at the max supported level
      * of the redpanda kafka server.
-     *
-     * TODO: this will go away once the kafka client implements version
-     * negotiation and can track on a per-broker basis the supported version
-     * range.
      */
     template<typename T>
     requires(KafkaApi<typename T::api_type>)
       ss::future<typename T::api_type::response_type> dispatch(T r) {
-        using type = std::remove_reference_t<std::decay_t<T>>;
-        if constexpr (std::is_same_v<type, offset_fetch_request>) {
-            return dispatch(std::move(r), api_version(4));
-        } else if constexpr (std::is_same_v<type, fetch_request>) {
-            return dispatch(std::move(r), api_version(10));
-        } else if constexpr (std::is_same_v<type, list_offsets_request>) {
-            return dispatch(std::move(r), api_version(3));
-        } else if constexpr (std::is_same_v<type, produce_request>) {
-            return dispatch(std::move(r), api_version(7));
-        } else if constexpr (std::is_same_v<type, offset_commit_request>) {
-            return dispatch(std::move(r), api_version(7));
-        } else if constexpr (std::is_same_v<type, describe_groups_request>) {
-            return dispatch(std::move(r), api_version(2));
-        } else if constexpr (std::is_same_v<type, heartbeat_request>) {
-            return dispatch(std::move(r), api_version(3));
-        } else if constexpr (std::is_same_v<type, join_group_request>) {
-            return dispatch(std::move(r), api_version(4));
-        } else if constexpr (std::is_same_v<type, sync_group_request>) {
-            return dispatch(std::move(r), api_version(3));
-        } else if constexpr (std::is_same_v<type, leave_group_request>) {
-            return dispatch(std::move(r), api_version(2));
-        } else if constexpr (std::is_same_v<type, metadata_request>) {
-            return dispatch(std::move(r), api_version(7));
-        } else if constexpr (std::is_same_v<type, find_coordinator_request>) {
-            return dispatch(std::move(r), api_version(2));
-        } else if constexpr (std::is_same_v<type, list_groups_request>) {
-            return dispatch(std::move(r), api_version(2));
-        } else if constexpr (std::is_same_v<type, create_topics_request>) {
-            return dispatch(std::move(r), api_version(4));
-        } else if constexpr (std::is_same_v<type, sasl_handshake_request>) {
-            return dispatch(std::move(r), api_version(1));
-        } else if constexpr (std::is_same_v<type, sasl_authenticate_request>) {
-            return dispatch(std::move(r), api_version(1));
+        if (!_max_versions) {
+            /// If request fails, this throws
+            _max_versions = co_await get_api_versions();
         }
+        if (
+          T::api_type::key() < 0
+          || T::api_type::key() > (_max_versions->size() - 1)) {
```

Comment:
This should probably be an assertion. Redpanda client uses api_keys from redpanda proper. So if redpanda returns invalid api keys to a redpanda client... its redpanda not working correctly